### ISSUE TITLE
[ONNX] Dynamo export freezes dynamic time dimension in manual bidirectional LSTM/GRU pattern

### ISSUE THREAD

**Issue by SouperDuper** [pytorch/pytorch#191855]
### 🐛 Describe the bug

`torch.onnx.export(..., dynamo=True)` exports a manual bidirectional LSTM or GRU pattern with a dynamic time axis, but the generated ONNX graph contains the example input time length in an internal `Expand` shape.

The exported model runs in ONNX Runtime for the example length, then fails for a different valid time length.

The same pattern works with `torch.onnx.export(..., dynamo=False)`.

The control cases in the full reproducer show that these related cases work with `dynamo=True`:

- bare unidirectional `nn.LSTM`
- bare unidirectional `nn.GRU`
- masked unidirectional `nn.LSTM`
- masked unidirectional `nn.GRU`

The clean reproduction environment used the CUDA 13.0 PyTorch wheel, but the reproducer itself executes entirely on CPU and uses ONNX Runtime's `CPUExecutionProvider`.

### Reproduction

Full reproducer (script + requirements.txt): https://gist.github.com/SouperDuper/d00d4e94e4461f9fc3c08bf7815d32fe

The full reproducer above imports only PyTorch, ONNX, ONNX Script, ONNX Runtime, NumPy, and standard-library modules. It adds LSTM and GRU control cases, compares the legacy exporter, and inspects the generated graph for literal Expand dimensions.

The slightly smaller reproducer below shows narrower failing case plus the `dynamo=False` comparison.

Run: `python filename.py`

```python
from __future__ import annotations

import tempfile
import logging
from contextlib import redirect_stderr, redirect_stdout
from io import StringIO
from pathlib import Path

import numpy as np
import onnxruntime as ort
import torch


logging.disable(logging.CRITICAL)

def reverse_valid_prefix(
    x: torch.Tensor,
    key_padding_mask: torch.Tensor,
) -> torch.Tensor:
    lengths = (~key_padding_mask).to(dtype=torch.long).sum(dim=1)
    positions = torch.ones_like(key_padding_mask, dtype=torch.long).cumsum(dim=1) - 1
    reversed_positions = lengths.unsqueeze(1) - 1 - positions
    source_positions = torch.where(
        positions < lengths.unsqueeze(1),
        reversed_positions,
        positions,
    ).clamp_min(0)
    return x.gather(1, source_positions.unsqueeze(-1).expand_as(x))


class ManualBidirectionalLstm(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.forward_rnn = torch.nn.LSTM(
            input_size=16,
            hidden_size=8,
            num_layers=1,
            batch_first=True,
            bidirectional=False,
            dropout=0.0,
        )
        self.backward_rnn = torch.nn.LSTM(
            input_size=16,
            hidden_size=8,
            num_layers=1,
            batch_first=True,
            bidirectional=False,
            dropout=0.0,
        )
        self.output = torch.nn.Linear(16, 16)

    def forward(self, x: torch.Tensor, key_padding_mask: torch.Tensor) -> torch.Tensor:
        forward, _ = self.forward_rnn(x)
        backward_input = reverse_valid_prefix(x, key_padding_mask)
        backward_reversed, _ = self.backward_rnn(backward_input)
        backward = reverse_valid_prefix(backward_reversed, key_padding_mask)
        return self.output(torch.cat((forward, backward), dim=-1))


def export_model(path: Path, *, dynamo: bool) -> None:
    model = ManualBidirectionalLstm().eval()
    x = torch.randn((1, 6, 16), dtype=torch.float32)
    key_padding_mask = torch.zeros((1, 6), dtype=torch.bool)

    kwargs = {
        "dynamo": dynamo,
        "external_data": False,
        "input_names": ["x", "key_padding_mask"],
        "output_names": ["y"],
        "opset_version": 18,
    }
    if dynamo:
        kwargs["dynamic_shapes"] = {
            "x": {0: torch.export.Dim.AUTO, 1: torch.export.Dim.AUTO},
            "key_padding_mask": {0: torch.export.Dim.AUTO, 1: torch.export.Dim.AUTO},
        }
    else:
        kwargs["dynamic_axes"] = {
            "x": {0: "batch", 1: "time"},
            "key_padding_mask": {0: "batch", 1: "time"},
            "y": {0: "batch", 1: "time"},
        }

    with redirect_stdout(StringIO()), redirect_stderr(StringIO()):
        torch.onnx.export(model, (x, key_padding_mask), path, **kwargs)


def run_onnx(path: Path, *, n_frames: int) -> tuple[bool, str]:
    session_options = ort.SessionOptions()
    session_options.log_severity_level = 4
    session = ort.InferenceSession(
        path.as_posix(),
        sess_options=session_options,
        providers=["CPUExecutionProvider"],
    )
    feeds = {
        "x": np.random.randn(1, n_frames, 16).astype(np.float32),
        "key_padding_mask": np.zeros((1, n_frames), dtype=np.bool_),
    }
    try:
        session.run(None, feeds)
    except Exception as exc:
        return False, str(exc).splitlines()[0]
    return True, ""


def main() -> None:
    torch.manual_seed(1234)
    output_dir = Path(tempfile.mkdtemp(prefix="onnx-dynamo-bidir-rnn-"))

    dynamo_path = output_dir / "manual_bidirectional_lstm.dynamo.onnx"
    export_model(dynamo_path, dynamo=True)
    dynamo_t6_ok, dynamo_t6_error = run_onnx(dynamo_path, n_frames=6)
    dynamo_t9_ok, dynamo_t9_error = run_onnx(dynamo_path, n_frames=9)

    legacy_path = output_dir / "manual_bidirectional_lstm.legacy.onnx"
    export_model(legacy_path, dynamo=False)
    legacy_t9_ok, legacy_t9_error = run_onnx(legacy_path, n_frames=9)

    print(f"dynamo_t6={'ok' if dynamo_t6_ok else 'fail'} error={dynamo_t6_error!r}")
    print(f"dynamo_t9={'ok' if dynamo_t9_ok else 'fail'} error={dynamo_t9_error!r}")
    print(f"legacy_t9={'ok' if legacy_t9_ok else 'fail'} error={legacy_t9_error!r}")


if __name__ == "__main__":
    main()
```

### Actual output

Narrower reproducer:

```text
dynamo_t6=ok error=''
dynamo_t9=fail error="[ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Non-zero status code returned while running Expand node. Name:'node_expand_1' Status Message: invalid expand shape"
legacy_t9=ok error=''
```

Full reproducer:

```text
output_dir=<tempdir>
python=3.11.14
torch=2.12.1+cu130
onnx=1.22.0
onnxscript=0.7.1
onnxruntime=1.28.0
numpy=1.26.4
lstm_bare: dynamo_export=ok dynamo_ort_t6=ok dynamo_ort_t9=ok error=''
lstm_masked_unidirectional: dynamo_export=ok dynamo_ort_t6=ok dynamo_ort_t9=ok error=''
lstm_manual_bidirectional: dynamo_export=ok dynamo_ort_t6=ok dynamo_ort_t9=fail error="[ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Non-zero status code returned while running Expand node. Name:'node_expand_1' Status Message: invalid expand shape"
lstm_manual_bidirectional: legacy_export=ok legacy_ort_t9=ok error='' dynamo_expand_shape_literals=[[1, 8], [16], [6, 8]]
gru_bare: dynamo_export=ok dynamo_ort_t6=ok dynamo_ort_t9=ok error=''
gru_masked_unidirectional: dynamo_export=ok dynamo_ort_t6=ok dynamo_ort_t9=ok error=''
gru_manual_bidirectional: dynamo_export=ok dynamo_ort_t6=ok dynamo_ort_t9=fail error="[ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Non-zero status code returned while running Expand node. Name:'node_expand_1' Status Message: invalid expand shape"
gru_manual_bidirectional: legacy_export=ok legacy_ort_t9=ok error='' dynamo_expand_shape_literals=[[1, 8], [16], [6, 8]]
```

The model is exported with an example input of shape `[1, 6, 16]`.

The repro only validates dynamic time behavior. It does not claim dynamic batch behavior, because the example batch size and test batch size are both `1`.

The public input time axis accepts a different length in the working cases. In the failing manual bidirectional cases from the full reproducer, the ONNX graph still contains the literal `6` in an internal `Expand` shape:

```text
dynamo_expand_shape_literals=[[1, 8], [16], [6, 8]]
```

This makes the exported model invalid for a different time length, such as `[1, 9, 16]`.

### Expected behavior

If `dynamo=True` export succeeds with a dynamic time axis, the exported ONNX model should run with a different valid sequence length.

The manual bidirectional LSTM and GRU cases should behave like the passing cases:

- export from `[1, 6, 16]`
- run in ONNX Runtime with `[1, 6, 16]`
- run in ONNX Runtime with `[1, 9, 16]`

### Why this looks like an exporter bug

The failing model is ordinary PyTorch code:

- two unidirectional `nn.LSTM` or `nn.GRU` modules
- `torch.gather` with a mask-derived reverse index
- `torch.cat`
- `torch.nn.Linear`

The same bidirectional model works with `torch.onnx.export(..., dynamo=False)`.

The bare RNN and masked unidirectional cases in the full reproducer also work with `dynamo=True`.

The failure is not ONNX Runtime rejecting the public input shape. ONNX Runtime accepts the dynamic public input, then fails inside the graph because a shape derived from the export example length is frozen into an internal `Expand`.

> **AI-assistance disclosure:** I do not have experience writing proper bug reports, so I used ChatGPT to help package and organize this report. I personally wrote the original failure case, ran the reproducer, reviewed the code and exported graphs, and verified the technical claims and reported output. AI was a sanity check, but it tends to make the resulting report too wordy.

### Versions

Collecting environment information...
PyTorch version: 2.12.1+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 IoT Enterprise LTSC (10.0.26100 64-bit)
GCC version: (MinGW-W64 x86_64-ucrt-posix-seh, built by Brecht Sanders, r8) 13.2.0

Clang version: Could not collect
CMake version: version 3.29.2
Libc version: N/A

Python version: 3.11.14 | packaged by conda-forge | (main, Oct 22 2025, 22:35:28) [MSC v.1944 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.26100-SP0
Is CUDA available: True
CUDA runtime version: 12.6.20

CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 596.49
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Name: AMD Ryzen 7 5800XT 8-Core Processor            
Manufacturer: AuthenticAMD
Family: 107
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 3801
MaxClockSpeed: 3801
L2CacheSize: 4096
L2CacheSpeed: None
Revision: 8450

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] onnx==1.22.0
[pip3] onnx-ir==0.2.1
[pip3] onnxruntime==1.28.0
[pip3] onnxscript==0.7.1
[pip3] torch==2.12.1+cu130
[conda] Could not collect

cc @justinchuby @titaiwangms
